test/test_deployment.py

```diff
@@ -0,0 +1,682 @@
+                        TASKS_IMAGE, 'webhook'],
+                       check=True)
+
+    # wait until RabbitMQ initialized
+    exec_c(data.tasks, f'until bots/inspect-queue --amqp {AMQP_POD}; do sleep 1; done', timeout=30)
+
+    yield data
+
+    subprocess.run(['podman', 'pod', 'rm', '-f', data.pod], check=True)
+
+
+#
+# Utilities
+#
+
+def exec_c(container: str, command: str, timeout=None, *, capture: bool = False,
+           exec_opts: list[str] | None = None) -> str | None:
+    """Run shell command in a container
+
+    Assert that it succeeds.
+    Return stdout if capture is True. stderr is not captured.
+
+    Default timeout is 5s.
+    """
+    res = subprocess.run(
+        ['podman', 'exec', '-i', *(exec_opts or []), container, 'sh', '-ec', command],
+        check=True,
+        stdout=subprocess.PIPE if capture else None,
+        timeout=5 if timeout is None else timeout)
+
+    return res.stdout.decode() if capture else None
+
+
+def exec_c_out(container: str, command: str, timeout=None) -> str:
+    """Run shell command in a container
+
+    Assert that it succeeds.
+    Return stdout. stderr is not captured.
+
+    Default timeout is 5s.
+    """
+    out = exec_c(container, command, timeout, capture=True)
+    assert out is not None  # mypy
+    return out
+
+
+def get_s3(config: Config, pod: PodData, path: str) -> str:
+    """Return the content of an S3 object"""
+
+    s3_base = f'https://localhost.localdomain:{pod.host_port_s3}'
+
+    context = ssl.create_default_context(cafile=config.secrets / 'ca.pem')
+    with urllib.request.urlopen(f'{s3_base}/{path}', context=context) as f:
+        return f.read().decode()
+
+
+#
+# Per test fixtures
+#
+
+@pytest.fixture()
+def bots_sha(pod: PodData) -> str:
+    """Return the SHA of the bots checkout"""
+
+    return exec_c_out(pod.tasks, 'git -C bots rev-parse HEAD').strip()
+
+
+@pytest.fixture()
+def mock_github(pod: PodData, bots_sha: str) -> Generator[str, None, None]:
+    """Start mock GitHub API server
+
+    Return environment shell command for using it
+    """
+    subprocess.run(['podman', 'cp', str(ROOT_DIR / 'tasks/mock-github'), f'{pod.tasks}:/work/bots/mock-github'],
+                   check=True)
+    mock_pid = exec_c_out(pod.tasks, f'''
+        cd bots
+        PYTHONPATH=. nohup ./mock-github --log /tmp/mock.log cockpit-project/bots {bots_sha} \
+                </dev/null > /tmp/mock.out 2>&1 &
+        echo $!''').strip()
+    # wait until it started
+    exec_c_out(pod.tasks, f'curl --retry 5 --retry-connrefused --fail {GHAPI_URL_POD}/repos/cockpit-project/bots')
+
+    yield f'export GITHUB_API={GHAPI_URL_POD}; SHA={bots_sha}'
+
+    exec_c_out(pod.tasks, f'while kill {mock_pid}; do sleep 0.1; done')
+
+
+@pytest.fixture()
+def mock_git_push(pod: PodData) -> Generator[None, None, None]:
+    """Install tasks/mock-git-push"""
+
+    subprocess.run(['podman', 'cp', str(ROOT_DIR / 'tasks/mock-git-push'), f'{pod.tasks}:/usr/local/bin/git'],
```

Comment:
I wonder if we can just include this in the pod setup?  Is there ever a time we want to push "for real"?

The interaction of one fixture implicitly modifying the state of another fixture (and then returning None for itself) seems a bit magic to me.

Reply:
Right, moved to the setup.